src/data/hostUrlSingleton.ts

```diff
@@ -0,0 +1,89 @@
+import type { ConfigurationParameters as IndexOperationsApiConfigurationParameters } from '../pinecone-generated-ts-fetch';
+import {
+  IndexOperationsApi,
+  Configuration as ApiConfiguration,
+} from '../pinecone-generated-ts-fetch';
+import type { PineconeConfiguration } from './types';
+import { middleware } from '../utils/middleware';
+import { queryParamsStringify, buildUserAgent, getFetch } from '../utils';
+import { describeIndex } from '../control';
+import { PineconeUnableToResolveHostError } from '../errors';
+
+// We use describeIndex to retrieve the data plane url (host) for a given API key
+// and index. We only ever want to call describeIndex a maximum of once per API key
+// and index so we cache them in a singleton for reuse.
+export const HostUrlSingleton = (function () {
+  const hostUrls = {}; // map of apiKey-indexName to hostUrl
+  let indexOperationsApi: InstanceType<typeof IndexOperationsApi> | null = null;
+
+  const _buildIndexOperationsApi = (config: PineconeConfiguration) => {
+    const { apiKey } = config;
+    const controllerPath = `https://api.pinecone.io`;
```

Comment:
Some way of overriding this value using an env var like `PINECONE_CONTROLLER_HOST` would be useful in getting the integration tests working over the short-term while api.pinecone.io is still being worked on.

Reply:
Added. I almost felt like we could leave `PINECONE_ENVIRONMENT` as an optional env variable for now and then either use `PINECONE_CONTROLLER_HOST` if provided, or `PINECONE_ENVIRONMENT` to build it if provided, or default to `https://api.pinecone.io`, but that seemed like it would overly complicate things when we can just provide the URL anyways.